howso/input_validation.amlg

```diff
@@ -441,113 +441,173 @@
 								)
 
 								(if (= (null) given_value)
-									(or
-										;not required
-										(or
-											;required not specified
-											(not (contains_index specification "required"))
-											;or its specifically false (this should never really be the case)
-											(= (get specification "required") (false))
-										)
-										;exp_type is null, or contains null
-										(= exp_type "null")
-										(and
-											(~ (list) exp_type)
-											(contains_value exp_type "null")
+									(if (and
+											(get specification "required")
+											(not (contains_value exp_type "null"))
 										)
+										[{ message "This value must be specified."}]
+										[]
 									)
+
 									(and
 										(~ (assoc) specification)
 										(get specification "required")
 										(= (null) given_value)
 									)
-									(false)
+									[{ message "This value must be specified."}]
 
 									;any_of logic
 									(contains_index specification "any_of")
-									(apply "or"
-										(map
-											(lambda
-												(call !SingleTypeCheck (append
-													{given_value given_value}
-													(if (~ (assoc) (current_value))
-														(assoc
-															specification (current_value 1)
-															exp_type (get (current_value 1) "type")
-														)
+									(if (apply "or"
+											(map
+												(lambda
+													(= (size
+														(call !SingleTypeCheck (append
+															{given_value given_value}
+															(if (~ (assoc) (current_value))
+																(assoc
+																	specification (current_value 1)
+																	exp_type (get (current_value 1) "type")
+																)
 
-														;otherwise, it should just be a string
-														(assoc
-															specification (assoc)
-															exp_type (current_value 1)
-														)
-													)
-												))
+																;otherwise, it should just be a string
+																(assoc
+																	specification (assoc)
+																	exp_type (current_value 1)
+																)
+															)
+														))
+													) 0)
+												)
+												(get specification "any_of")
 											)
-											(get specification "any_of")
 										)
-									)
+										[]
 
+										[{ message "The given value does not match any of the types specified under \"any_of\"."}]
```

Comment:
Not sure this is helpful to a user seeing the message. They wont know what "any_of" means